x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/CreateApiKeyResponse.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.action;
+
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.time.Instant;
+
+/**
+ * Response for the successful creation of an api key
+ */
+public final class CreateApiKeyResponse extends ActionResponse {
+
+    private String name;
+    private String key;
```

Comment:
I made the switch to secure string but I am not really sure it is worth it. Technically the key will be a String when looking up the document. @tvernum wdyt?

Reply:
It is a valid concern. We could model this more after what github does where they use this as the password for the user in basic auth

Reply:
I think this is a real concern.
The current model stores unhashed credentials in an index, and we shouldn't be doing that.

I think there's a few options:
- As suggested above, require the userid with in the auth header. Then we can store api-keys bcrypt'ed , and to authenticated, we find all api-key docs for the specified user, and check the provided key against each hash.
- Use separate "access key" and "secret key" like AWS (although I hate those names - "token id" and "token password" have clearer meanings). The token-id is used for the doc id, and the token-password is bcrypt'ed in the doc (this is similar to the option above, but each token has an id, which is included in the auth header, rather than just using the userid)
- As above, but combine those parts into a single string that is handled as a single private credential. So you get an "api key" of `467b45d3:a87f991c3ed6cada` which logically two parts, but we mostly handle it as a single string.
 - Require signing of API-key requests (I briefly discussed this with @robgil in Dublin). The API key is more of less public information, and it has a public key associated with it. Each request to the API must be signed by the corresponding private key. This pushes all the problem to standard crypto which is nice, but forces all usage of api keys to support RSA, and storing private keys, etc.

Reply:
@jkakavas would be interested to hear your thoughts if you have any

Reply:
The api-key is in essence a potentially long lived Bearer token so we should be treating this similarly to how we'd treat a password. (there are differences with regards to ease of revocation and rotation, and smaller user impact as it is not likely to be reused in other services like a password, etc. but still )
That means that we should not be storing them in plaintext in an index. The rest of the doc doesn't contain any information that we should care about protecting somehow. The obvious first choice is to store a hash of the api-key and compare to that on authentication. I agree with the first three suggestions that @tvernum shared above with a few comments:
- We don't need to use a salt. Assuming that the api-key is long enough and with enough entropy there is no threat of a pre-computed hash table attack. SHA256 would be ok for now, but we probably should use an adaptive function (PBKDF2 or bcrypt) for the sake of configuration only future-proofing against brute force attacks.
- I started of leaning towards the first suggestion as it forces the client to keep a track of who it gets tokens for and simplifies our implementation but do we also verify that the user exists and still has the roles on every execution or do we use the userid just as a id to search for api-key docs ? If the latter is true, then the api-key-id + api-key solution makes more sense. 
- Signing of API-Key requests feels like pushing the problem one turtle down as then the clients need to store private keys and we need to store their public keys and we need to allow them to register their public keys before use, which sounds complicated for the use case at hand.

Reply:
> we probably should use an adaptive function (PBKDF2 or bcrypt) for the sake of configuration only future-proofing against brute force attacks.

Agreed. I am going to go with PBKDF2 so its good to go in terms of FIPS.

> do we also verify that the user exists and still has the roles on every execution or do we use the userid just as a id to search for api-key docs

Userid would only be used to search for api-key docs at this time. *maybe if we do shadow users* we could look up info about the user but we shouldn't design for something we don't have a great deal of certainty about adding.

My proposal for moving forward is that we:

* generate a key id + key
* store a pbkdf2 hash of the key. the client has the key and we discard it
* use key id to lookup doc and then use pbkdf2 to validate the key; we'll probably want to cache using a faster hash like we do with regular passwords

Reply:
@jkakavas I'm not sure what you're referring to with `Signing API-Key requests`. That sounds like cryptographically signing an api key, which is not what I was proposing. To clarify what I was asking about in Dublin was specifically `request signing` with the intention of thwarting replay attacks. It might involve adding a TTL header. With a signed request, the TTL could not be changed by an attacker without invalidating the signed request. It might be better to move `request signing` to a separate issue/feature discussion.